### PR TITLE
[Bridges] Fix ConstraintIndex conflicts between variable and constraint bridges

### PR DESCRIPTION
**Pull request by blegat**
Closes https://github.com/jump-dev/MathOptInterface.jl/pull/2358

### CODE REVIEW

**Comment by odow** [comment]
Let me know when you get this to a point that you want me to take a look.

**Comment by blegat** [comment]
Ready for review

File: src/Bridges/Constraint/map.jl [hunk 1]
```diff
@@ -274,19 +274,39 @@ end
...
 function add_key_for_bridge end
 
+_ensure_available(::Map, ::Type, ::Type, ::Function) = nothing
+function _ensure_available(
+    map::Map,
+    F::Type{MOI.VectorOfVariables},
+    ::Type{S},
+    is_available::Function,
+) where {S}
+    while !is_available(MOI.ConstraintIndex{F,S}(-length(map.bridges) - 1))
```
Comment (odow) [inline src/Bridges/Constraint/map.jl:294]: So this is a bit confusing. We're just going to push in a bunch of `nothing` until the index becomes available?

What if this ends up adding thousands of elements?

File: src/Bridges/Variable/map.jl [hunk 2]
```diff
@@ -338,26 +382,42 @@ function add_key_for_bridge(
...
     push!(map.parent_index, map.current_context)
     bridge_index = Int64(length(map.parent_index))
-    push!(map.info, -MOI.dimension(set))
+    F = MOI.VectorOfVariables
+    S = typeof(set)
+    while !is_available(
+        MOI.ConstraintIndex{F,S}(-length(map.vector_of_variables_map) - 1),
+    )
+        push!(map.vector_of_variables_map, 0)
+        push!(map.vector_of_variables_length, 0)
```
Comment (odow) [inline src/Bridges/Variable/map.jl:416]: This needs tests

File: src/Bridges/Variable/map.jl [hunk 3]
```diff
@@ -338,26 +382,42 @@ function add_key_for_bridge(
...
-`typeof(set)`. It returns a tuple with `variables` and the constraint index
-`MOI.ConstraintIndex{MOI.VectorOfVariables, typeof(set)}(first(variables).value)`.
+`typeof(set)`. It returns a tuple with `variables` and a constraint index
+`ci::MOI.ConstraintIndex{MOI.VectorOfVariables, typeof(set)}` such that
+`is_available(ci)`.
 """
 function add_keys_for_bridge(
     map::Map,
     bridge_fun::Function,
     set::MOI.AbstractVectorSet,
```
Comment (odow) [inline src/Bridges/Variable/map.jl:401]:
```suggestion
    set::S,
```

File: src/Bridges/Variable/map.jl [hunk 4]
```diff
@@ -338,26 +382,42 @@ function add_key_for_bridge(
...
+    is_available::Function,
 )
```
Comment (odow) [inline src/Bridges/Variable/map.jl:403]:
```suggestion
) where {S<:MOI.AbstractVectorSet}
```

File: src/Bridges/Variable/map.jl [hunk 5]
```diff
@@ -338,26 +382,42 @@ function add_key_for_bridge(
...
     if iszero(MOI.dimension(set))
         return MOI.VariableIndex[],
         MOI.ConstraintIndex{MOI.VectorOfVariables,typeof(set)}(0)
     end
     push!(map.parent_index, map.current_context)
     bridge_index = Int64(length(map.parent_index))
-    push!(map.info, -MOI.dimension(set))
+    F = MOI.VectorOfVariables
+    S = typeof(set)
```
Comment (odow) [inline src/Bridges/Variable/map.jl:411]:
```suggestion
```

**Review by odow — COMMENTED** [review]
I feel like this needs far more tests. I always get a bit nervous with these changes.

https://github.com/jump-dev/MathOptInterface.jl/actions/runs/7268307681

**Comment by odow** [comment]
This looks like it broke a bunch of solver tests, including Gurobi and CPLEX. So as well as the fix, we obviously need more tests in MOI proper.

File: src/Bridges/Constraint/map.jl (hunk 1 above)
Comment (blegat) [reply src/Bridges/Constraint/map.jl:294]: It means you also have thousands of constrained vector of variables. Another solution would be to have one common vector shared between `Constraint.Map` and `Variable.Map`

Comment (blegat) [reply src/Bridges/Constraint/map.jl:294]: It just kind of break encapsulation so it's a tradeoff

**Comment by blegat** [comment]
> we obviously need more tests in MOI proper.

Yes, that's also the motivation for https://github.com/jump-dev/MathOptInterface.jl/pull/2357

**Comment by blegat** [comment]
New run: https://github.com/jump-dev/MathOptInterface.jl/actions/runs/7273606634

**Comment by blegat** [comment]
New run: https://github.com/jump-dev/MathOptInterface.jl/actions/runs/7277998597

File: src/Bridges/Constraint/map.jl (hunk 1 above)
Comment (odow) [reply src/Bridges/Constraint/map.jl:294]: Fair enough. Let's wait until it's a problem.

**Comment by odow** [comment]
I don't really have an opinion on the design. But I also don't have a suggestion for anything better.

I do think we need 100% coverage before merging though.

**Comment by blegat** [comment]
> I do think we need 100% coverage before merging though.

We do now

**Review by odow — APPROVED** [review]
I'll merge so we can re-trigger #2357.

I'll also pay careful attention to `solver-tests` before tagging a new MOI release.

**Pull request merged**